chain/chain/src/stateless_validation/spice_chunk_validation.rs

```diff
@@ -220,3 +220,719 @@ fn validate_source_receipts_proofs(
+    use crate::{BlockProcessingArtifact, Provenance};
+
+    use super::*;
+
+    const TEST_VALIDATORS: [&str; 2] = ["test-validator-1", "test-validator-2"];
+
+    #[test]
+    #[cfg_attr(not(feature = "protocol_feature_spice"), ignore)]
+    fn test_pre_validation_output_is_correct() {
+        let test_chain = setup();
+        let witness = test_chain.valid_witness();
+
+        let output = test_chain.run_pre_validation(&witness).unwrap();
+        assert!(output.implicit_transition_params.is_empty());
+        let MainTransition::NewChunk { new_chunk_data, block_hash: _ } =
+            output.main_transition_params
+        else {
+            panic!()
+        };
+
+        let prev_execution_results = test_chain.prev_execution_results();
+        let prev_chunk_header = test_chain.prev_chunk_header();
+        let spice_chunk_header = test_chain.chunk_header().into_spice_chunk_execution_header(
+            &prev_execution_results.0.get(prev_chunk_header.chunk_hash()).unwrap().chunk_extra,
+        );
+        assert_eq!(new_chunk_data.chunk_header, spice_chunk_header);
+
+        let receipts = test_chain.receipts_for_shard(prev_chunk_header.shard_id());
+        assert_eq!(new_chunk_data.receipts, receipts);
+
+        let transactions = test_chain.transactions();
+        let new_chunk_data_transactions =
+            new_chunk_data.transactions.into_nonexpired_transactions();
+        assert_eq!(new_chunk_data_transactions, transactions);
+    }
+
+    #[test]
+    #[cfg_attr(not(feature = "protocol_feature_spice"), ignore)]
+    fn test_pre_validation_fails_with_incorrect_epoch_id() {
+        let test_chain = setup();
+        let valid_witness = test_chain.valid_witness();
+
+        let invalid_epoch_id =
+            EpochId::from_str("32222222222233333333334444444444445555555777").unwrap();
+        let invalid_witness = ChunkStateWitness::new(
+            AccountId::from_str("unused").unwrap(),
+            invalid_epoch_id,
+            valid_witness.chunk_header().clone(),
+            valid_witness.main_state_transition().clone(),
+            valid_witness.source_receipt_proofs().clone(),
+            *valid_witness.applied_receipts_hash(),
+            valid_witness.transactions().clone(),
+            valid_witness.implicit_transitions().clone(),
+            valid_witness.new_transactions().clone(),
+            PROTOCOL_VERSION,
+        );
+
+        let error_message = unwrap_error_message(test_chain.run_pre_validation(&invalid_witness));
+        assert_contains(&error_message, "does not match epoch id");
+    }
+
+    #[test]
+    #[cfg_attr(not(feature = "protocol_feature_spice"), ignore)]
+    fn test_pre_validation_fails_with_unrelated_chunk_header() {
+        let test_chain = setup();
+        let valid_witness = test_chain.valid_witness();
+
+        let invalid_chunk_header = test_chain.prev_chunk_header();
+        let invalid_witness = ChunkStateWitness::new(
```

Comment:
it would be nice to avoid this verbose pattern to copy most of the field from the old witness.
how about adding helper method here, something like `update_witness` that takes a witness and a function that modifies `ChunkStateWitnessV2`, so the test would look something like:
```
let invalid_witness = update_witness(valid_witness, |v2| v2.chunk_header = invalid_chunk_header);
```